// xqtt/lib/packets/connect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _helpers = require('./helpers');

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

exports.default = {
  encode: function encode(packet) {
    var packetType = 1;
    var flags = 0;

    var protocolName = (0, _helpers.encodeUTF8String)('MQTT');
    var protocolLevel = 4;

    var usernameFlag = !!packet.username;
    var passwordFlag = !!packet.password;
    var willRetain = !!(packet.will && packet.will.retain);
    var willQoS = packet.will && packet.will.qos || 0;
    var willFlag = !!packet.will;
    var cleanSession = packet.clean || typeof packet.clean === 'undefined';
    var connectFlags = (usernameFlag ? 128 : 0) + (passwordFlag ? 64 : 0) + (willRetain ? 32 : 0) + (willQoS & 2 ? 16 : 0) + (willQoS & 1 ? 8 : 0) + (willFlag ? 4 : 0) + (cleanSession ? 2 : 0);

    var keepAlive = packet.keepAlive && typeof packet.keepAlive !== 'undefined' ? packet.keepAlive : 0;

    var variableHeader = [].concat(_toConsumableArray(protocolName), [protocolLevel, connectFlags, keepAlive >> 8, keepAlive & 0xff]);

    var payload = [].concat(_toConsumableArray((0, _helpers.encodeUTF8String)(packet.clientId)));

    if (packet.username) {
      payload.push.apply(payload, _toConsumableArray((0, _helpers.encodeUTF8String)(packet.username)));
    }

    if (packet.password) {
      payload.push.apply(payload, _toConsumableArray((0, _helpers.encodeUTF8String)(packet.password)));
    }

    var fixedHeader = [packetType << 4 | flags].concat(_toConsumableArray((0, _helpers.encodeLength)(variableHeader.length + payload.length)));

    return [].concat(_toConsumableArray(fixedHeader), _toConsumableArray(variableHeader), _toConsumableArray(payload));
  },
  decode: function decode(buffer, remainingLength) {
    var protocolNameStart = buffer.length - remainingLength;
    var protocolName = (0, _helpers.decodeUTF8String)(buffer, protocolNameStart);

    var protocolLevelIndex = protocolNameStart + protocolName.length;
    var protocolLevel = buffer[protocolLevelIndex];

    var connectFlagsIndex = protocolLevelIndex + 1;
    var connectFlags = buffer[connectFlagsIndex];
    var usernameFlag = !!(connectFlags & 128);
    var passwordFlag = !!(connectFlags & 64);
    var willRetain = !!(connectFlags & 32);
    var willQoS = (connectFlags & 16 + 8) >> 3;
    var willFlag = !!(connectFlags & 4);
    var cleanSession = !!(connectFlags & 2);

    if (willQoS !== 0 && willQoS !== 1 && willQoS !== 2) {
      throw new Error('invalid will qos');
    }

    var keepAliveIndex = connectFlagsIndex + 1;
    var keepAlive = (buffer[keepAliveIndex] << 8) + buffer[keepAliveIndex + 1];

    var clientIdStart = keepAliveIndex + 2;
    var clientId = (0, _helpers.decodeUTF8String)(buffer, clientIdStart);

    var username = void 0;
    var password = void 0;

    var usernameStart = clientIdStart + clientId.length;

    if (usernameFlag) {
      username = (0, _helpers.decodeUTF8String)(buffer, usernameStart);
    }

    if (passwordFlag) {
      var passwordStart = usernameStart + (username ? username.length : 0);
      password = (0, _helpers.decodeUTF8String)(buffer, passwordStart);
    }

    return {
      type: 'connect',
      protocolName: protocolName.value,
      protocolLevel: protocolLevel,
      clientId: clientId.value,
      username: username ? username.value : undefined,
      password: password ? password.value : undefined,
      will: willFlag ? {
        retain: willRetain,
        qos: willQoS
      } : undefined,
      clean: cleanSession,
      keepAlive: keepAlive
    };
  }
};
