// xqtt/lib/packets/helpers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.encodeLength = encodeLength;
exports.decodeLength = decodeLength;
exports.encodeUTF8String = encodeUTF8String;
exports.toUTF8Array = toUTF8Array;
exports.decodeUTF8String = decodeUTF8String;

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function encodeLength(x) {
  var output = [];

  do {
    var encodedByte = x % 128;

    x = Math.floor(x / 128);

    if (x > 0) {
      encodedByte = encodedByte | 128;
    }

    output.push(encodedByte);
  } while (x > 0);

  return output;
}

function decodeLength(buffer, startIndex) {
  var i = startIndex;
  var encodedByte = 0;
  var value = 0;
  var multiplier = 1;

  do {
    encodedByte = buffer[i++];

    value += (encodedByte & 127) * multiplier;

    multiplier *= 128;

    if (multiplier > 128 * 128 * 128) {
      throw Error('malformed length');
    }
  } while ((encodedByte & 128) != 0);

  return { length: value, bytesUsedToEncodeLength: i - startIndex };
}

function encodeUTF8String(str) {
  var bytes = toUTF8Array(str);

  return [bytes.length >> 8, bytes.length & 0xff].concat(_toConsumableArray(bytes));
}

// https://stackoverflow.com/a/18729931
function toUTF8Array(str) {
  var utf8 = [];
  for (var i = 0; i < str.length; i++) {
    var charcode = str.charCodeAt(i);
    if (charcode < 0x80) utf8.push(charcode);else if (charcode < 0x800) {
      utf8.push(0xc0 | charcode >> 6, 0x80 | charcode & 0x3f);
    } else if (charcode < 0xd800 || charcode >= 0xe000) {
      utf8.push(0xe0 | charcode >> 12, 0x80 | charcode >> 6 & 0x3f, 0x80 | charcode & 0x3f);
    } else {
      // surrogate pair
      i++;
      // UTF-16 encodes 0x10000-0x10FFFF by
      // subtracting 0x10000 and splitting the
      // 20 bits of 0x0-0xFFFFF into two halves
      charcode = 0x10000 + ((charcode & 0x3ff) << 10 | str.charCodeAt(i) & 0x3ff);
      utf8.push(0xf0 | charcode >> 18, 0x80 | charcode >> 12 & 0x3f, 0x80 | charcode >> 6 & 0x3f, 0x80 | charcode & 0x3f);
    }
  }
  return utf8;
}

function decodeUTF8String(buffer, startIndex) {
  var length = (buffer[startIndex] << 8) + buffer[startIndex + 1];
  var bytes = buffer.slice(startIndex + 2, startIndex + 2 + length);
  var value = Utf8ArrayToStr(bytes);

  return {
    length: length + 2,
    value: value
  };
}

// https://stackoverflow.com/a/22373135
// http://www.onicos.com/staff/iz/amuse/javascript/expert/utf.txt

/* utf.js - UTF-8 <=> UTF-16 convertion
 *
 * Copyright (C) 1999 Masanao Izumo <iz@onicos.co.jp>
 * Version: 1.0
 * LastModified: Dec 25 1999
 * This library is free.  You can redistribute it and/or modify it.
 */

function Utf8ArrayToStr(array) {
  var out, i, len, c;
  var char2, char3;

  out = '';
  len = array.length;
  i = 0;
  while (i < len) {
    c = array[i++];
    switch (c >> 4) {
      case 0:
      case 1:
      case 2:
      case 3:
      case 4:
      case 5:
      case 6:
      case 7:
        // 0xxxxxxx
        out += String.fromCharCode(c);
        break;
      case 12:
      case 13:
        // 110x xxxx   10xx xxxx
        char2 = array[i++];
        out += String.fromCharCode((c & 0x1f) << 6 | char2 & 0x3f);
        break;
      case 14:
        // 1110 xxxx  10xx xxxx  10xx xxxx
        char2 = array[i++];
        char3 = array[i++];
        out += String.fromCharCode((c & 0x0f) << 12 | (char2 & 0x3f) << 6 | (char3 & 0x3f) << 0);
        break;
    }
  }
  return out;
}
