// xqtt/lib/packets/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.encode = encode;
exports.decode = decode;

var _helpers = require('./helpers');

var _connect = require('./connect');

var _connect2 = _interopRequireDefault(_connect);

var _connack = require('./connack');

var _connack2 = _interopRequireDefault(_connack);

var _publish = require('./publish');

var _publish2 = _interopRequireDefault(_publish);

var _puback = require('./puback');

var _puback2 = _interopRequireDefault(_puback);

var _pubrec = require('./pubrec');

var _pubrec2 = _interopRequireDefault(_pubrec);

var _pubrel = require('./pubrel');

var _pubrel2 = _interopRequireDefault(_pubrel);

var _pubcomp = require('./pubcomp');

var _pubcomp2 = _interopRequireDefault(_pubcomp);

var _subscribe = require('./subscribe');

var _subscribe2 = _interopRequireDefault(_subscribe);

var _suback = require('./suback');

var _suback2 = _interopRequireDefault(_suback);

var _unsubscribe = require('./unsubscribe');

var _unsubscribe2 = _interopRequireDefault(_unsubscribe);

var _unsuback = require('./unsuback');

var _unsuback2 = _interopRequireDefault(_unsuback);

var _pingreq = require('./pingreq');

var _pingreq2 = _interopRequireDefault(_pingreq);

var _pingres = require('./pingres');

var _pingres2 = _interopRequireDefault(_pingres);

var _disconnect = require('./disconnect');

var _disconnect2 = _interopRequireDefault(_disconnect);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var packetTypesByName = {
  connect: _connect2.default,
  connack: _connack2.default,
  publish: _publish2.default,
  puback: _puback2.default,
  pubrec: _pubrec2.default,
  pubrel: _pubrel2.default,
  pubcomp: _pubcomp2.default,
  subscribe: _subscribe2.default,
  suback: _suback2.default,
  unsubscribe: _unsubscribe2.default,
  unsuback: _unsuback2.default,
  pingreq: _pingreq2.default,
  pingres: _pingres2.default,
  disconnect: _disconnect2.default
};

var packetTypesById = [null, _connect2.default, // 1
_connack2.default, // 2
_publish2.default, // 3
_puback2.default, // 4
_pubrec2.default, // 5
_pubrel2.default, // 6
_pubcomp2.default, // 7
_subscribe2.default, // 8
_suback2.default, // 9
_unsubscribe2.default, // 10
_unsuback2.default, // 11
_pingreq2.default, // 12
_pingres2.default, // 13
_disconnect2.default];

function encode(packet) {
  var name = packet.type;
  var packetType = packetTypesByName[name];

  if (!packetType) {
    throw new Error('packet type ' + name + ' cannot be encoded');
  }

  return packetType.encode(packet);
}

function decode(buffer) {
  if (buffer.length < 2) {
    return null;
  }

  var id = buffer[0] >> 4;
  var packetType = packetTypesById[id];

  if (!packetType) {
    throw new Error('packet type ' + id + ' cannot be decoded');
  }

  var _decodeLength = (0, _helpers.decodeLength)(buffer, 1),
      remainingLength = _decodeLength.length,
      bytesUsedToEncodeLength = _decodeLength.bytesUsedToEncodeLength;

  if (buffer.length < 1 + bytesUsedToEncodeLength + remainingLength) {
    return null;
  }

  return packetType.decode(buffer, remainingLength);
}
