// xqtt/lib/packets/publish.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _helpers = require('./helpers');

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

exports.default = {
  encode: function encode(packet) {
    var packetType = 3;

    var qos = packet.qos || 0;

    var flags = (packet.dup ? 8 : 0) + (qos & 2 ? 4 : 0) + (qos & 1 ? 2 : 0) + (packet.retain ? 1 : 0);

    var variableHeader = [].concat(_toConsumableArray((0, _helpers.encodeUTF8String)(packet.topic)));

    if (qos === 1 || qos === 2) {
      if (typeof packet.id !== 'number' || packet.id < 1) {
        throw new Error('when qos is 1 or 2, packet must have id');
      }

      variableHeader.push(packet.id >> 8, packet.id & 0xff);
    }

    var payload = packet.payload;

    if (typeof payload === 'string') {
      payload = (0, _helpers.toUTF8Array)(payload);
    }

    var fixedHeader = [packetType << 4 | flags].concat(_toConsumableArray((0, _helpers.encodeLength)(variableHeader.length + payload.length)));

    return [].concat(_toConsumableArray(fixedHeader), _toConsumableArray(variableHeader), _toConsumableArray(payload));
  },
  decode: function decode(buffer, remainingLength) {
    var flags = buffer[0] & 0x0f;

    var dup = !!(flags & 8);
    var qos = (flags & 6) >> 1;
    var retain = !!(flags & 1);

    if (qos !== 0 && qos !== 1 && qos !== 2) {
      throw new Error('invalid qos');
    }

    var topicStart = buffer.length - remainingLength;
    var decodedTopic = (0, _helpers.decodeUTF8String)(buffer, topicStart);
    var topic = decodedTopic.value;

    var id = 0;
    var payloadStart = topicStart + decodedTopic.length;

    if (qos > 0) {
      var idStart = payloadStart;

      id = (buffer[idStart] << 8) + buffer[idStart + 1];

      payloadStart += 2;
    }

    var payload = buffer.slice(payloadStart);

    return {
      type: 'publish',
      topic: topic,
      payload: payload,
      dup: dup,
      retain: retain,
      qos: qos,
      id: id
    };
  }
};
