// xqtt/lib/client.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _log = require('./log');

var _log2 = _interopRequireDefault(_log);

var _packets = require('./packets');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var packetIdLimit = Math.pow(2, 16);

var Client = function () {
  function Client(options) {
    _classCallCheck(this, Client);

    this.options = options || {};
    this.clientId = this.options.clientId || this.generateClientId();
    this.keepAlive = this.options.keepAlive || this.defaultKeepAlive;
    this.connectionState = 'never-connected';
    this.reconnectAttempt = 0;
    this.lastPacketId = 0;
  }

  // Public methods

  _createClass(Client, [{
    key: 'connect',
    value: function connect(reconnecting) {
      var _this = this;

      switch (this.connectionState) {
        case 'never-connected':
        case 'connect-failed':
        case 'offline':
        case 'disconnected':
          break;
        default:
          this.log('should not be connecting in ' + this.connectionState + ' state');
          return;
      }

      this.changeState(reconnecting ? 'reconnecting' : 'connecting');

      return new Promise(function (resolve, reject) {
        _this.resolveConnect = resolve;
        _this.rejectConnect = reject;

        _this.open();
      });
    }
  }, {
    key: 'publish',
    value: function publish(topic, payload, options) {
      switch (this.connectionState) {
        case 'connected':
          break;
        default:
          this.log('should not be publishing in ' + this.connectionState + ' state');
          return;
      }

      var qos = options && options.qos || 0;
      var id = qos > 0 ? this.nextPacketId() : 0;

      this.send({
        type: 'publish',
        dup: options && options.dup || false,
        retain: options && options.retain || false,
        topic: topic,
        payload: payload,
        qos: qos,
        id: id
      });

      // TODO: if qos > 0, message needs to be ack'ed
    }
  }, {
    key: 'subscribe',
    value: function subscribe(topic, qos) {
      switch (this.connectionState) {
        case 'connected':
          break;
        default:
          this.log('should not be subscribing in ' + this.connectionState + ' state');
          return;
      }

      this.send({
        type: 'subscribe',
        id: this.nextPacketId(),
        subscriptions: [{ topic: topic, qos: qos || 0 }]
      });
    }
  }, {
    key: 'unsubscribe',
    value: function unsubscribe(topic) {
      switch (this.connectionState) {
        case 'connected':
          break;
        default:
          this.log('should not be unsubscribing in ' + this.connectionState + ' state');
          return;
      }

      this.send({
        type: 'unsubscribe',
        id: this.nextPacketId(),
        topics: [topic]
      });
    }
  }, {
    key: 'disconnect',
    value: function disconnect() {
      if (this.connectTimer) {
        clearTimeout(this.connectTimer);
        this.connectTimer = null;
      }

      if (this.reconnectTimer) {
        clearTimeout(this.reconnectTimer);
        this.reconnectTimer = null;
      }

      if (this.keepAliveTimer) {
        clearTimeout(this.keepAliveTimer);
        this.keepAliveTimer = null;
      }

      switch (this.connectionState) {
        case 'connected':
          break;
        default:
          this.log('should not be disconnecting in ' + this.connectionState + ' state');
          return;
      }

      this.changeState('disconnecting');

      this.send({ type: 'disconnect' });

      this.close();
    }

    // Connection methods implemented by subclasses, consider abtstract and protected

  }, {
    key: 'open',
    value: function open() {
      throw new Error('not implemented');
    }
  }, {
    key: 'write',
    value: function write(_bytes) {
      throw new Error('not implemented');
    }
  }, {
    key: 'close',
    value: function close() {
      throw new Error('not implemented');
    }

    // Connection methods invoked by subclasses, consider protected

  }, {
    key: 'connectionOpened',
    value: function connectionOpened() {
      // this.log('connectionOpened');

      this.startConnectTimer();

      this.send({
        type: 'connect',
        clientId: this.clientId,
        keepAlive: this.keepAlive,
        username: this.options.username,
        password: this.options.password
      });
    }
  }, {
    key: 'connectionClosed',
    value: function connectionClosed() {
      // this.log('connectionClosed');

      switch (this.connectionState) {
        case 'disconnecting':
          this.changeState('disconnected');
          break;
        case 'connected':
          this.changeState('offline');
          this.reconnectAttempt = 0;
          this.startReconnectTimer();
          break;
        case 'connecting':
        case 'reconnecting':
          if (this.connectionState === 'connecting') {
            this.reconnectAttempt = 0;
          } else {
            this.reconnectAttempt++;
          }

          this.changeState('connect-failed');
          this.startReconnectTimer();

          break;
        default:
          this.log('connection should not be closing in ' + this.connectionState + ' state');
          break;
      }
    }
  }, {
    key: 'connectionError',
    value: function connectionError(error) {
      // TODO: decide what to do with this
      this.log('connectionError', error);
    }
  }, {
    key: 'bytesReceived',
    value: function bytesReceived(buffer) {
      this.emit('bytesreceived', buffer);

      var bytes = this.incomingBuffer ? Uint8Array.from([].concat(_toConsumableArray(this.incomingBuffer), _toConsumableArray(buffer))) : buffer;

      var packet = this.decode(bytes);

      if (packet) {
        this.packetReceived(packet);

        this.incomingBuffer = null;
      } else {
        this.incomingBuffer = bytes;
      }
    }

    // Methods that can be overridden by subclasses, consider protected

  }, {
    key: 'packetReceived',
    value: function packetReceived(packet) {
      this.emit('packetreceive', packet);

      switch (packet.type) {
        case 'connack':
          this.handleConnack();
          break;
        case 'publish':
          this.handlePublish(packet);
          break;
        case 'puback':
          this.handlePuback(packet);
          break;
        case 'pubrec':
          this.handlePubrec(packet);
          break;
        case 'pubrel':
          this.handlePubrel(packet);
          break;
        case 'pubcomp':
          this.handlePubcomp(packet);
          break;
        case 'suback':
          this.handleSuback(packet);
          break;
        case 'unsuback':
          this.handleUnsuback(packet);
          break;
      }
    }
  }, {
    key: 'protocolViolation',
    value: function protocolViolation(msg) {
      this.log('protocolViolation', msg);
    }
  }, {
    key: 'handleConnack',
    value: function handleConnack() {
      switch (this.connectionState) {
        case 'connecting':
        case 'reconnecting':
          break;
        default:
          this.log('should not be receiving connack packets in ' + this.connectionState + ' state');
          return;
      }

      var wasConnecting = this.connectionState === 'connecting';

      this.changeState('connected');

      if (wasConnecting && this.resolveConnect) {
        this.resolveConnect(this);
      }

      this.startKeepAliveTimer();
    }
  }, {
    key: 'handlePublish',
    value: function handlePublish(packet) {
      if (packet.qos === 1) {
        if (typeof packet.id !== 'number' || packet.id < 1) {
          return this.protocolViolation('publish packet with qos 1 is missing id');
        }

        this.send({
          type: 'puback',
          id: packet.id
        });
      } else if (packet.qos === 2) {
        if (typeof packet.id !== 'number' || packet.id < 1) {
          return this.protocolViolation('publish packet with qos 2 is missing id');
        }

        this.send({
          type: 'pubrec',
          id: packet.id
        });
      }
    }
  }, {
    key: 'handlePuback',
    value: function handlePuback(_packet) {
      // TODO: mark message as acknowledged
    }
  }, {
    key: 'handlePubrec',
    value: function handlePubrec(packet) {
      // TODO: mark message as received
      this.send({
        type: 'pubrel',
        id: packet.id
      });
    }
  }, {
    key: 'handlePubrel',
    value: function handlePubrel(packet) {
      // TODO: mark message as released
      this.send({
        type: 'pubcomp',
        id: packet.id
      });
    }
  }, {
    key: 'handlePubcomp',
    value: function handlePubcomp(_packet) {
      // TODO: mark message as completely acknowledged
    }
  }, {
    key: 'handleSuback',
    value: function handleSuback(_packet) {
      // TODO: mark subscription as acknowledged
    }
  }, {
    key: 'handleUnsuback',
    value: function handleUnsuback(_packet) {
      // TODO: mark unsubscription as acknowledged
    }
  }, {
    key: 'startConnectTimer',
    value: function startConnectTimer() {
      var _this2 = this;

      this.connectTimer = setTimeout(function () {
        if (_this2.connectionState !== 'connected') {
          var wasConnecting = _this2.connectionState === 'connecting';

          _this2.changeState('connect-failed');

          if (wasConnecting) {
            _this2.reconnectAttempt = 0;

            _this2.rejectConnect(new Error('connect timed out'));
          }

          _this2.startReconnectTimer();
        } else {
          _this2.log('connectTimer should have been cancelled');
        }
      }, this.options.connectTimeout || this.defaultConnectTimeout);
    }
  }, {
    key: 'startKeepAliveTimer',
    value: function startKeepAliveTimer() {
      var _this3 = this;

      var elapsed = new Date() - this.lastPacketTime;
      var timeout = this.keepAlive * 1000 - elapsed;

      this.keepAliveTimer = setTimeout(function () {
        return _this3.sendKeepAlive();
      }, timeout);
    }
  }, {
    key: 'sendKeepAlive',
    value: function sendKeepAlive() {
      if (this.connectionState === 'connected') {
        var elapsed = new Date() - this.lastPacketTime;
        var timeout = this.keepAlive * 1000;

        if (elapsed >= timeout) {
          this.send({
            type: 'pingreq'
          });
        }

        this.startKeepAliveTimer();
      } else {
        this.log('keepAliveTimer should have been cancelled');
      }
    }
  }, {
    key: 'startReconnectTimer',
    value: function startReconnectTimer() {
      var _this4 = this;

      var options = this.options;

      if (options.reconnect === false) {
        return;
      }

      var defaultReconnectOptions = this.defaultReconnectOptions;

      var reconnectOptions = _typeof(options.reconnect) === 'object' ? options.reconnect : defaultReconnectOptions;

      // https://github.com/tim-kos/node-retry
      // https://dthain.blogspot.com/2009/02/exponential-backoff-in-distributed.html
      var random = 1 + (reconnectOptions.random ? Math.random() : 0);
      var factor = reconnectOptions.factor || defaultReconnectOptions.factor;
      var minTimeout = reconnectOptions.minTimeout || defaultReconnectOptions.minTimeout;
      var maxTimeout = reconnectOptions.maxTimeout || defaultReconnectOptions.maxTimeout;
      var attempt = this.reconnectAttempt;
      var delay = Math.min(random * minTimeout * Math.pow(factor, attempt), maxTimeout);

      this.log('reconnecting in ' + delay + 'ms');

      this.reconnectTimer = setTimeout(function () {
        _this4.connect(true);
      }, delay);
    }

    // Utility methods

  }, {
    key: 'changeState',
    value: function changeState(newState) {
      var oldState = this.connectionState;

      this.connectionState = newState;

      this.emit('statechange', { from: oldState, to: newState });

      this.emit(newState);
    }
  }, {
    key: 'generateClientId',
    value: function generateClientId() {
      var prefix = this.options.clientIdPrefix || this.defaultClientIdPrefix;
      var suffix = Math.random().toString(36).slice(2);

      return prefix + '-' + suffix;
    }
  }, {
    key: 'nextPacketId',
    value: function nextPacketId() {
      this.lastPacketId = (this.lastPacketId + 1) % packetIdLimit;

      // Don't allow packet id to be 0.
      if (!this.lastPacketId) {
        this.lastPacketId = 1;
      }

      return this.lastPacketId;
    }
  }, {
    key: 'send',
    value: function send(packet) {
      this.emit('packetsend', packet);

      var bytes = this.encode(packet);

      this.emit('bytessent', bytes);

      this.write(bytes);

      this.lastPacketTime = new Date();
    }
  }, {
    key: 'encode',
    value: function encode(packet) {
      return (0, _packets.encode)(packet);
    }
  }, {
    key: 'decode',
    value: function decode(bytes) {
      return (0, _packets.decode)(bytes);
    }
  }, {
    key: 'emit',
    value: function emit(event, data) {
      if (typeof this.options[event] === 'function') {
        this.options[event](data);
      }
    }
  }]);

  return Client;
}();

exports.default = Client;


Client.prototype.defaultClientIdPrefix = 'xqtt';
Client.prototype.defaultConnectTimeout = 30 * 1000;
Client.prototype.defaultKeepAlive = 45;
Client.prototype.defaultReconnectOptions = {
  random: true,
  factor: 2,
  minTimeout: 1000,
  maxTimeout: Infinity
};
Client.prototype.log = _log2.default;
