// nqtt/lib/client.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _net = require('net');

var _net2 = _interopRequireDefault(_net);

var _xqtt = require('xqtt');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Client extends _xqtt.Client {

  constructor(options) {
    super(options);

    this.handleConnectEvent = () => {
      this.connectionOpened();
    };

    this.handleCloseEvent = () => {
      this.connectionClosed();
    };

    this.handleErrorEvent = e => {
      this.connectionError(e);
    };

    this.handleDataEvent = data => {
      const buffer = new Uint8Array(data);

      this.bytesReceived(buffer);
    };
  }

  open() {
    const { host, port } = this.options;

    this.socket = _net2.default.connect(port, host);

    this.socket.on('connect', this.handleConnectEvent);
    this.socket.on('close', this.handleCloseEvent);
    this.socket.on('error', this.handleErrorEvent);
    this.socket.on('data', this.handleDataEvent);
  }

  write(bytes) {
    const buffer = Buffer.from(bytes);

    this.socket.write(buffer);
  }

  close() {
    this.socket.end();
  }

}

exports.default = Client;
Client.prototype.defaultClientIdPrefix = 'nqtt';
